{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genbank phylogeny from COI sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Complete mitogenome genbank records were downloaded for 4 ant species, each from a different subfamily:\n",
    "  - ***Pseudomyrmex gracilis*** (Pseudomyrmecinae)\n",
    "  - ***Formica fusca*** (Formicinae)\n",
    "  - ***Linepthema humile*** (Dolichoderinae)\n",
    "  - ***Solenopsis invicta*** (Myrmicinae)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing everything we'll need\n",
    "from Bio import SeqIO\n",
    "import os, glob\n",
    "import pandas as pd\n",
    "import skbio.io"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we'll be using SeqIO to obtain the COI nucleotide sequence for each species and save it into separate files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_dir(dir_name):\n",
    "    os.makedirs(os.path.dirname(dir_name), exist_ok=True)\n",
    "\n",
    "def extract_COI(gb_file):\n",
    "    for record in SeqIO.parse(gb_file, \"genbank\"):\n",
    "        species_name = record.annotations.get('organism').replace(\" \", \"_\")\n",
    "        filename = \"./coi_seqs/{}_coi.fa\".format(species_name)\n",
    "        create_dir(filename)\n",
    "        with open(filename, \"w\") as coi_file:\n",
    "            for gene in record.features:\n",
    "                if gene.type in [\"CDS\"] and gene.qualifiers.get('gene')[0] in ['COX1', 'COI']:\n",
    "                    header = \"{}-{}\".format(species_name, gene.qualifiers.get('gene')[0])\n",
    "                    sequence = gene.location.extract(record.seq) # Mas seq pode ter stop codon truncado\n",
    "                    if len(sequence) % 3 == 1:\n",
    "                        sequence += \"AA\" #Resto 1 - Precisa adicionar 'AA'\n",
    "                    elif len(sequence) % 3 == 2:\n",
    "                        sequence += \"A\" #Resto 2 - Precisa adicionar 'A'\n",
    "                    coi_file.write(\">{}\\n{}\\n\".format(header, sequence))\n",
    "\n",
    "for gb_file in glob.glob(\"./ant_mitogenomes/*.gb\"):\n",
    "    extract_COI(gb_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running blastn with the COI sequences against NCBI's formicidae sequences:\n",
    "\n",
    "**OBS:** Need to install [taxdb database](ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxdb dir found\n",
      "Setting BLASTDB variable to /home/gabriel/Dropbox/repos/genbank_phylogeny/taxdb\n",
      "blast_results directory already created\n",
      "Running blast search for ./coi_seqs/Formica_fusca_coi.fa...\n",
      "Running blast search for ./coi_seqs/Linepithema_humile_coi.fa...\n",
      "Running blast search for ./coi_seqs/Pseudomyrmex_gracilis_coi.fa...\n",
      "Running blast search for ./coi_seqs/Solenopsis_invicta_coi.fa...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Extracting taxdb database (if necessary)\n",
    "taxdb=$(pwd)/taxdb\n",
    "if [[ ! -d \"$taxdb\" ]]; then \n",
    "    echo \"taxdb dir not found. Extracting taxdb database...\"\n",
    "    mkdir $taxdb\n",
    "    tar -C $taxdb -xaf taxdb.tar.gz\n",
    "else\n",
    "    echo \"taxdb dir found\"\n",
    "fi\n",
    "\n",
    "# Setting BLASTDB variable\n",
    "echo \"Setting BLASTDB variable to $taxdb\"\n",
    "export BLASTDB=$taxdb\n",
    "\n",
    "# Creating directory for blast results (if necessary)\n",
    "if [[ ! -d \"blast_results\" ]]; then \n",
    "    echo \"Creating dirrectory for blast results\"\n",
    "    mkdir blast_results\n",
    "else\n",
    "    echo \"blast_results directory already created\"\n",
    "fi\n",
    "\n",
    "# Performing blast searches \n",
    "for coi in ./coi_seqs/*; do\n",
    "    echo \"Running blast search for $coi...\" && \n",
    "    #blastn -query $coi -db ./blast_teste/ant_mito -out ./blast_results/$(basename $coi .fa).blast -outfmt \"6 qseqid sseqid staxids sscinames stitle sacc saccver slen sstart send qseq\";\n",
    "    blastn -query $coi -db nr -max_target_seqs 100 -remote -entrez_query \"Formicidae [Organism]\" -outfmt \"7 qseqid sseqid staxids sscinames stitle sacc saccver slen sstart send qseq\" -out ./blast_results/$(basename $coi .fa).blast; # set max_target_seqs to higher value when running this for real\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NOTE:** Blast with the `-remote` flag can take quite some time to run and is not compatible with `-taxidlist`... If hard disk space is not a problem, maybe it would be better to just download the entire nt database and run it locally..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Saving the blast results into dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gabriel/miniconda3/lib/python3.8/site-packages/skbio/io/registry.py:547: FormatIdentificationWarning: <_io.TextIOWrapper name='blast_results/Formica_fusca_coi_old.blast' mode='r' encoding='UTF-8'> does not look like a blast+7 file\n",
      "  warn(\"%r does not look like a %s file\"\n",
      "<ipython-input-2-37e3d7676a29>:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['sseq'] = df['sseq'].apply(lambda x: x.replace('-', ''))\n",
      "<ipython-input-2-37e3d7676a29>:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['sseqlen'] = df.apply(lambda row: len(row.sseq), axis = 1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 qseqid                staxids  \\\n",
      "0    Formica_fusca-COX1                  72779   \n",
      "1    Formica_fusca-COX1                 208979   \n",
      "2    Formica_fusca-COX1                  72779   \n",
      "3    Formica_fusca-COX1                  72779   \n",
      "4    Formica_fusca-COX1                  84560   \n",
      "..                  ...                    ...   \n",
      "495  Formica_fusca-COX1                 609739   \n",
      "496  Formica_fusca-COX1                 609832   \n",
      "497  Formica_fusca-COX1                 609832   \n",
      "498  Formica_fusca-COX1                  72779   \n",
      "499  Formica_fusca-COX1  609832;609890;2328943   \n",
      "\n",
      "                                             sscinames      sacc  sseqlen  \\\n",
      "0                                        Formica fusca  LN607805     1530   \n",
      "1                                       Formica selysi  KP670862     1530   \n",
      "2                                        Formica fusca  FJ824419     1377   \n",
      "3                                        Formica fusca  AY334398     1191   \n",
      "4                                       Formica lemani  AB019425      974   \n",
      "..                                                 ...       ...      ...   \n",
      "495                                    Formica accreta  LT977373      658   \n",
      "496                              Formica neorufibarbis  KR895865      658   \n",
      "497                              Formica neorufibarbis  KR933364      658   \n",
      "498                                      Formica fusca  JF863560      658   \n",
      "499  Formica neorufibarbis;Formica subaenescens;For...  HM374821      658   \n",
      "\n",
      "                                                  sseq  \n",
      "0    ATGAATAAATGATTATATTCTACAAATCATAAAGATATTGGTATTC...  \n",
      "1    ATGAATAAATGATTATATTCTACAAATCATAAAGATATTGGTATTC...  \n",
      "2    ATAATTATTCGCTTAGAATTAGGCTCATCAAATTCATTAATCAATA...  \n",
      "3    AATTTTATTAATGACGGAACAGGAACAGGATGAACTATTTATCCTC...  \n",
      "4    ATTCCCTTAATACTAGGATCTCCAGACATAGCTTATCCTCGTATAA...  \n",
      "..                                                 ...  \n",
      "495  TATTCTTTACTTCTTATTTGCTATTTGAGCAGGAATAATTGGATCT...  \n",
      "496  TATTCTTTACTTCTTATTTGCTATTTGAGCAGGAATAATTGGATCT...  \n",
      "497  TATTCTTTACTTCTTATTTGCTATTTGAGCAGGAATAATTGGATCT...  \n",
      "498  TATTCTTTACTTCTTATTTGCTATTTGAGCAGGAATAATTGGATCT...  \n",
      "499  TATTCTTTACTTCTTATTTGCTATTTGAGCAGGAATAATTGGATCT...  \n",
      "\n",
      "[500 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "def create_dataframe(blast_result):\n",
    "    with open(blast_result) as blast7:\n",
    "        df = skbio.io.read(blast7, format='blast+7', into=pd.DataFrame)\n",
    "        return df\n",
    "\n",
    "def filter_dataframe(df):\n",
    "    df = df[['qseqid', 'staxids', 'sscinames', 'sacc', 'sseq']]\n",
    "    df['sseq'] = df['sseq'].apply(lambda x: x.replace('-', ''))\n",
    "    #df['sseqlen'] = len(df['sseq'])\n",
    "    df['sseqlen'] = df.apply(lambda row: len(row.sseq), axis = 1) \n",
    "    df = df[['qseqid', 'staxids', 'sscinames', 'sacc', 'sseqlen', 'sseq']]\n",
    "    return df\n",
    "    \n",
    "df = filter_dataframe(create_dataframe('blast_results/Formica_fusca_coi_old.blast'))\n",
    "print(df)\n",
    "df.to_excel(\"blast.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([8, 306, 314, 499], dtype='int64')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Testing methods of dataframe\n",
    "df[df[\"staxids\"].str.contains(\";\")].index\n",
    "#df.index\n",
    "#df[\";\" in df.staxids].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        sscinames       sacc  staxids  sseqlen\n",
      "164                               Messor bouvieri   DQ074325    88063     1248\n",
      "109                         Oecophylla smaragdina   AB185475    84561     1042\n",
      "4                                  Formica lemani   AB019425    84560      974\n",
      "40                              Polyrhachis dives   KT266831    84555     1530\n",
      "168                               Messor structor   KT184578    81629     1367\n",
      "93                                Myrmica vandeli   GQ255194   758485     1371\n",
      "269                      Myrmica n. sp. 6 GJ-2010   GQ255162   758483     1376\n",
      "103                      Myrmica n. sp. 3 GJ-2010   GQ255191   758480     1371\n",
      "50                       Myrmica n. sp. 2 GJ-2010   GQ255164   758479     1379\n",
      "388                           Myrmica n. sp. M350   GQ255205   758476     1377\n",
      "419                           Myrmica n. sp. M295   GQ255201   758472     1377\n",
      "372                           Myrmica n. sp. M273   GQ255199   758470     1376\n",
      "431                          Proformica longiseta   AB010934    72795      968\n",
      "308                          Cataglyphis ibericus   DQ353343    72789     1029\n",
      "444                         Cataglyphis floricola   AB010936    72788      973\n",
      "457                       Cataglyphis rosenhaueri   AB010933    72787      973\n",
      "389                             Polyergus samurai   AB010930    72784      974\n",
      "9                               Formica truncorum   AB103360    72783     1041\n",
      "11                              Formica yessensis   AB103357    72782     1041\n",
      "7                                 Formica exsecta   AB103364    72781     1039\n",
      "62                            Formica cunicularia   AB010926    72780      974\n",
      "0                                   Formica fusca   LN607805    72779     1530\n",
      "429                    Paraparatrechina sp. PHL02   FJ982472   710416     1048\n",
      "470                          Nylanderia sp. GUY02   FJ982453   710409     1053\n",
      "474                          Nylanderia sp. GUY01   FJ982452   710408     1052\n",
      "456                          Nylanderia sp. FIJ01   FJ982451   710407     1053\n",
      "446                          Nylanderia sp. CHR01   FJ982446   710405     1046\n",
      "443                Nylanderia cf. pubens SGB-2010   FJ982462   710403     1053\n",
      "393       Nylanderia cf. madagascarensis SGB-2010   FJ982456   710402     1038\n",
      "55                          Myrmica punctiventris   GQ255168    68986     1377\n",
      "26                            Nylanderia flavipes  NC_049861    67766     1530\n",
      "352                        Wasmannia auropunctata   KX146469    64793     1530\n",
      "484                       Prenolepis melanogaster   JN562438   621474     1053\n",
      "479                       Phrynoponera gabonensis   KU504914   615737     1205\n",
      "463                               Pheidole hospes   KJ141815   615056     1133\n",
      "464                          Nylanderia vitiensis   FJ982465   614003     1053\n",
      "368                              Nylanderia otome   KY770018   613966     1194\n",
      "486                       Paraparatrechina glabra   FJ982469   613913     1051\n",
      "468                           Nylanderia amblyops   FJ982442   613864     1047\n",
      "84                          Odontomachus laticeps   KU504892   613448     1461\n",
      "340                              Myrmica wesmaeli   GQ255208   612875     1377\n",
      "144                               Myrmica tulinae   GQ255193   612870     1371\n",
      "258                         Myrmica transsibirica   GQ255189   612868     1377\n",
      "124                               Myrmica tobiasi   GQ255192   612867     1371\n",
      "335                                Myrmica salina   GQ255176   612849     1371\n",
      "112                   Myrmica rugulososcabrinodis   GQ255173   612846     1371\n",
      "355                          Myrmica lobulicornis   GQ255159   612822     1377\n",
      "96                                 Myrmica laurae   GQ255157   612820     1371\n",
      "66                                 Myrmica indica   GQ255147   612804     1379\n",
      "203                              Myrmica georgica   GQ255145   612799     1371\n",
      "259                             Myrmica divergens   GQ255137   612792     1377\n",
      "369                             Myrmica anatolica   GQ255126   612777     1377\n",
      "104                                 Myrmica aloba   GQ255187   612776     1371\n",
      "58                                Messor ibericus   KT184523   612033     1367\n",
      "343                               Huberia brounii   FJ824422   610398     1377\n",
      "472                          Formica subaenescens   MK037284   609890      681\n",
      "482                             Formica podzolica   LT977413   609855      658\n",
      "493                                 Formica picea   LT977411   609851      658\n",
      "496                         Formica neorufibarbis   KR895865   609832      658\n",
      "416                            Formica gagatoides   KX665068   609795      658\n",
      "367                               Formica cinerea   MG757150   609761      658\n",
      "488                                Formica aerata   JF863557   609743      658\n",
      "476                               Formica accreta   LT978303   609739      676\n",
      "440                          Euprenolepis procera   FJ982439   609683     1052\n",
      "51                            Anochetus targionii   KU504865   602530     1454\n",
      "90                            Anochetus subcoecus   KU504864   602527     1432\n",
      "79                               Anochetus simoni   KU504857   602522     1485\n",
      "143                          Anochetus obscuratus   KU504849   602502     1428\n",
      "35                            Anochetus neglectus   KU504848   602500     1460\n",
      "59                               Anochetus minans   KU504847   602494     1485\n",
      "44                          Anochetus altisquamis   KU504831   602461     1485\n",
      "140                              Acropyga smithii  NC_046399   602228     1530\n",
      "21                               Acropyga sauteri  NC_046398   602226     1530\n",
      "17                            Acropyga panamensis  NC_046426   602222     1530\n",
      "15                                 Acropyga myops  NC_046424   602216     1530\n",
      "29                             Acropyga kinomurai  NC_046423   602213     1525\n",
      "32                            Acropyga guianensis  NC_046422   602208     1530\n",
      "13                               Acropyga goeldii  NC_046420   602207     1530\n",
      "27                             Acropyga fuhrmanni  NC_046421   602205     1520\n",
      "480                           Pseudomyrmex peperi   BK010385   600674     1525\n",
      "87                               Myrmica wittmeri   GQ255196   571787     1379\n",
      "354                          Myrmica saposhnikovi   GQ255177   571781     1377\n",
      "326                           Myrmica rugiventris   GQ255171   571779     1371\n",
      "91                             Myrmica punctinops   GQ255167   571778     1377\n",
      "353                             Myrmica nearctica   GU109467   571777     1377\n",
      "312                            Myrmica lobicornis   FJ824437   571774     1377\n",
      "145                             Myrmica lacustris   GQ255156   571773     1371\n",
      "303                           Myrmica kirghisorum   GQ255153   571772     1377\n",
      "204                          Myrmica fracticornis   GQ255135   571769     1377\n",
      "292                              Myrmica formosae   FJ824433   571768     1377\n",
      "342                             Myrmica forcipata   GQ255140   571767     1371\n",
      "313                               Myrmica excelsa   FJ824432   571766     1377\n",
      "327                              Myrmica eidmanni   GQ255139   571765     1371\n",
      "186                           Myrmica dshungarica   GQ255138   571764     1379\n",
      "392                           Myrmica discontinua   GQ255136   571763     1377\n",
      "370                          Myrmica brevispinosa   FJ824431   571761     1365\n",
      "328                           Myrmica angulinodis   GQ255127   571758     1371\n",
      "48                       Myrmica aimonissabaudiae   GQ255122   571756     1370\n",
      "341                             Myrmica pisarskii   GQ255166   525332     1377\n",
      "115                                  Messor minor   EU441233   519485     1203\n",
      "339                                Atta laevigata   KC346251   493208     1527\n",
      "49                             Anochetus goodmani   KU504836   476159     1471\n",
      "283                Colobopsis cylindrica s.l. SAB   EF634190   450416     1239\n",
      "344                 Colobopsis cylindrica s.l. CH   EF634198   450408     1239\n",
      "351                            Camponotus sp. INT   EF634202   450398     1239\n",
      "31                       Odontomachus assiniensis   KU504878   443813     1465\n",
      "83             Lasius jensi x Lasius meridionalis   EF191429   437477     1215\n",
      "320                         Myrmica striolagaster   GQ255188   411767     1377\n",
      "212                         Camponotus conithorax   EF653271   411739     1239\n",
      "142                      Odontomachus troglodytes   KU504910   373794     1473\n",
      "492                         Pseudolasius typhlops   FJ982484   369227     1053\n",
      "454                        Prolasius sp. CSM-2006   DQ353353   369222     1034\n",
      "471                             Proformica nasuta   DQ353311   369220     1035\n",
      "448                      Plagiolepis sp. CSM-2006   DQ353352   369204     1035\n",
      "409                         Notoncus sp. CSM-2006   DQ353386   369181     1035\n",
      "331                  Myrmecorhynchus sp. CSM-2006   DQ353366   369174     1032\n",
      "5                                Formica wheeleri   DQ353362   369134     1035\n",
      "420                     Brachymyrmex sp. CSM-2006   DQ353294   369103     1035\n",
      "47                                Anochetus mayri   KU504845   369091     1482\n",
      "410                           Myrmecina nipponica   KX016626   357229     1243\n",
      "25                         Anoplolepis gracilipes  NC_039576   354296     1530\n",
      "23                          Acropyga donisthorpei   MH158404   354293     1530\n",
      "268               Messor cf. structor BCCS-2005-6   DQ074344   340073     1248\n",
      "345               Messor cf. structor BCCS-2005-5   DQ074340   340072     1248\n",
      "330               Messor cf. structor BCCS-2005-4   DQ074337   340069     1248\n",
      "336               Messor cf. structor BCCS-2005-3   DQ074338   340068     1248\n",
      "82                Messor cf. structor BCCS-2005-2   DQ074353   340067     1244\n",
      "106               Messor cf. structor BCCS-2005-1   DQ074345   337457     1248\n",
      "65                             Anochetus graeffei   KU504837   320728     1455\n",
      "475                           Pheidole capellinii   KJ141778   297324     1091\n",
      "435                           Pseudomyrmex viduus   KP271174   292178     1415\n",
      "319                       Pseudomyrmex dendroicus   BK010473   292155     1494\n",
      "88                            Lasius sp. FMS-2003   AY225878   283733     1202\n",
      "374                             Myrmica tahoensis   GQ255190    27493     1377\n",
      "485                       Myrmecocystus mexicanus   DQ353330   271132     1035\n",
      "231                     Anochetus madagascarensis   KU504842   269310     1418\n",
      "42                          Anochetus grandidieri   KU504838   269309     1484\n",
      "417                      Formica sp. BOLD:AAA4978   HM405608  2659124      658\n",
      "483                      Formica sp. BOLD:AAD8688   HQ928744  2659120      658\n",
      "359                          Formica paralugubris   EU600788   258705      776\n",
      "350                             Formica aquilonia   JX311439   258703      787\n",
      "299                            Cryptopone sauteri  NC_041202   255801     1526\n",
      "61                           Odontomachus rixosus   KU504905   255797     1464\n",
      "415                              Formica japonica   KX665091   255794      658\n",
      "56                         Aphaenogaster famelica  NC_049859   255788     1530\n",
      "315                         Camponotus schaefferi   AY334388   251257     1254\n",
      "382                               Camponotus sayi   AY334385   251256     1191\n",
      "380                           Camponotus ocreatus   AY334387   251254     1211\n",
      "405                         Camponotus americanus   AY334395   251250     1188\n",
      "53                               Messor mcarthuri   KT184517  2491882     1362\n",
      "122                                Messor muticus   KT184529  2491881     1367\n",
      "78                                Messor ponticus   KT184512  2491880     1367\n",
      "371                           Polyergus breviceps   DQ353325   239269     1032\n",
      "447                            Myopopone castanea   DQ353381   237740     1035\n",
      "146                               Lasius turcicus   DQ975427   235463     1170\n",
      "272                              Lasius sakagamii   AY225864   231991     1199\n",
      "133                           Lasius psammophilus   AY225863   231990     1199\n",
      "213                                 Lasius mixtus   AY225879   231988     1202\n",
      "77                               Lasius lasioides   AY225874   231987     1199\n",
      "247                            Lasius fuliginosus   EF191431   231986     1208\n",
      "271                            Lasius emarginatus   AY225868   231985     1199\n",
      "69                                Lasius brunneus   AY225877   231984     1199\n",
      "119                             Lasius austriacus   AY225872   231983     1199\n",
      "284                            Myrmica sulcinodis   AY280606   229918     1449\n",
      "85                               Myrmica sabuleti   AY280603   229917     1449\n",
      "52                               Myrmica rugulosa   AY280602   229916     1449\n",
      "152                            Myrmica microrubra   AY280598   229915     1449\n",
      "118                            Myrmica karavajevi   AY280596   229914     1449\n",
      "86                                Myrmica hirsuta   AY280595   229913     1449\n",
      "64                              Myrmica hellenica   AY280593   229912     1449\n",
      "6                               Formica pratensis   AB103363   221681     1041\n",
      "421                       Tetramorium bicarinatum   AY909379   219812     1115\n",
      "298                                 Manica rubida   AY280592   219785     1449\n",
      "1                                  Formica selysi   KP670862   208979     1530\n",
      "185                               Formica gerardi   MF276902   208976      816\n",
      "41                            Myrmica scabrinodis   LN607806   207696     1530\n",
      "95                               Ectatomma ruidum   MG870246   196312     1530\n",
      "491                       Crematogaster lineolata   KP252162   195111      665\n",
      "285                       Pseudoneoponera rufipes   KU504916  1885326     1463\n",
      "466              Leptogenys sp. USNM ENT 01124373   KU504873  1884783     1463\n",
      "46                Anochetus sp. USNM ENT 01124362   KU504858  1884775     1452\n",
      "473                          Myrmecina sp. 1HK_BG   KX021820  1828532     1276\n",
      "94                                   Formica moki   AF398151   168620      886\n",
      "489                          Pheidole sp. EPEM185   KJ141897  1580983     1140\n",
      "477                          Pheidole sp. EPEM085   KJ141803  1580937     1135\n",
      "478                          Pheidole sp. EMSM170   KJ141833  1580900     1134\n",
      "430                          Pheidole sp. EMSM107   KJ141797  1580899     1133\n",
      "366                          Oecophylla longinoda   AB185830   154973     1040\n",
      "436                                Zatania cisipa   JN562442  1177146     1054\n",
      "490                          Prenolepis sp. PHL02   JN562439  1177138     1050\n",
      "187                              Lasius neglectus   AY225875   111072     1199\n",
      "117                                 Myrmica rubra   AY280600   106198     1449\n",
      "113  Messor minor x Messor cf. wasmanni BCSS-2011   EU441259  1031672     1203\n",
      "73                  Messor cf. wasmanni BCSS-2011   KT184590  1031671     1366\n"
     ]
    }
   ],
   "source": [
    "def clean_dataframe(df):\n",
    "    clean_df = df.drop(df[df.staxids.str.contains(\";\")].index) # Removing rows with hybrid sequences (more than one taxid value)\n",
    "    clean_df = clean_df.sort_values(by=[\"staxids\", \"sseqlen\"], ascending=False) # Sorting dataframe by taxid and sseqlen (descending) - Guarantees that highest sseqlen will always be the first row for that taxid\n",
    "    # Printing all rows to check output\n",
    "    #with pd.option_context('display.max_rows', None, 'display.max_columns', None): \n",
    "    #    print(clean_df[[\"sscinames\", \"sacc\", \"staxids\", \"sseqlen\"]])\n",
    "    clean_df = clean_df.drop_duplicates(subset=\"staxids\", keep='first') # Keeps only one record per txid. The one that has the highest sseqlen\n",
    "    return clean_df\n",
    "\n",
    "with pd.option_context('display.max_rows', None, 'display.max_columns', None): \n",
    "    print(clean_dataframe(df)[[\"sscinames\", \"sacc\", \"staxids\", \"sseqlen\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
